File: dp/pro_4.cpp
#include <iostream>
#include <vector>
using namespace std;

int solution(vector<vector<int> > land)
{
    int DP[100000][4]={0,};
    int answer = 0;

    for(int i=0; i<4; i++){
        DP[0][i]=land[0][i];
    }
    for(int i=1; i<land.size(); i++){
        for(int j=0; j<4; j++){
            switch(j){
                case 0:
                    DP[i][j]=max(max(DP[i-1][1],DP[i-1][2]),DP[i-1][3])+land[i][j];
                    break;
                case 1:
                   DP[i][j]=max(max(DP[i-1][0],DP[i-1][2]),DP[i-1][3])+land[i][j];
                    break;
                case 2:
                    DP[i][j]=max(max(DP[i-1][1],DP[i-1][0]),DP[i-1][3])+land[i][j];
                    break;
                case 3:
                    DP[i][j]=max(max(DP[i-1][1],DP[i-1][2]),DP[i-1][0])+land[i][j];
                    break;
            }
        }
    }
    
    for(int i=0; i<4; i++){
        answer=max(max(DP[land.size()-1][0],DP[land.size()-1][1]),max(DP[land.size()-1][2],DP[land.size()-1][3]));
    }

    return answer;
}

File: 15th/pro_4.java
import java.util.PriorityQueue;
import java.util.Collections;

class Solution {
    public int[] solution(String[] operations) {
        int[] answer = new int[2];
        PriorityQueue<Integer> pq1 = new PriorityQueue<>();//min
        PriorityQueue<Integer> pq2 = new PriorityQueue<>(Collections.reverseOrder());//max
        for(int i=0; i<operations.length; i++){
            if(operations[i].charAt(0)=='I'){
                int temp = Integer.parseInt(operations[i].substring(2));
                pq1.add(temp);
                pq2.add(temp);
            }
            else{
                if(pq1.isEmpty())   continue;
                int rm;
                if(operations[i].charAt(2)=='1'){
                    rm = pq2.poll();
                    pq1.remove(rm);
                }else{
                    rm = pq1.poll();
                    pq2.remove(rm);
                }
            }
        }
        if(pq1.isEmpty()) return answer;
        else{
            answer[0]=pq2.peek();
            answer[1]=pq1.peek();
        }
        return answer;
    }
}